### PR TITLE
New Admin API client-side implementation

### PR DESCRIPTION
**Pull request by donatello**
This is a companion PR to minio/minio#5351 - please refer to it for the context. It includes client side changes for using the new Admin APIs.

### CODE REVIEW

File: cmd/admin-heal.go [hunk 1]
```diff
@@ -19,37 +19,41 @@ package cmd
...
+		Name:  "dry-run, n",
+		Usage: "Only inspect data, but do not mutate",
+	},
+	cli.BoolFlag{
+		Name:  "incomplete, I",
+		Usage: "Heal uploads recursively",
+	},
+	cli.BoolFlag{
+		Name:  "force-start, f",
+		Usage: "Force start a heal sequence",
```
Comment (vadmeste) [inline cmd/admin-heal.go]: I think this is better 'Force start a new heal sequence'

File: cmd/admin-heal.go [hunk 2]
```diff
@@ -72,304 +76,556 @@ EXAMPLES:
...
+	colRow := hColTable[parityShards]
+	for index, val := range colRow {
+		if val != -1 && surplusShards <= val {
+			return hColOrder[index], err
+		}
+	}
+	return c, fmt.Errorf("this will not be returned")
+}
+
+type hri struct {
```
Comment (vadmeste) [inline cmd/admin-heal.go]: Any reason why we declare this struct ? wouldn't be enough if we use madmin.HealResultItem ?

Comment (vadmeste) [reply cmd/admin-heal.go]: Can you comment this struct ?

File: cmd/admin-heal.go [hunk 3]
```diff
@@ -72,304 +76,556 @@ EXAMPLES:
...
+	if surplusShards < 0 {
+		return hColGrey, err
 	}
+	colRow := hColTable[parityShards]
+	for index, val := range colRow {
+		if val != -1 && surplusShards <= val {
+			return hColOrder[index], err
+		}
+	}
+	return c, fmt.Errorf("this will not be returned")
```
Comment (vadmeste) [inline cmd/admin-heal.go]: I think this is a better error msg: 'cannot get a heal color code'

**Comment by donatello** [comment]
> We need to provide proper errors when run against an FS mode, this error is not useful.
> 
> ```
> $ mc admin heal myminio
> mc: <ERROR> Failed to start heal sequence A header you provided implies functionality that is not implemented
> mc: <ERROR> Unable to display follow heal status A header you provided implies functionality that is not implemented
> 
> ```

The above is being addressed by adding a new server error via https://github.com/minio/minio/pull/5383

> 
> Are we still supporting?
> ```
> 
>   4. Heal all objects including uploads under 'dir' prefix
>        $ mc admin heal --incomplete --recursive play/testbucket/dir/
> 
> ```

Yes, adding `--incomplete` flag heals incomplete multipart uploads too.

> ```
> $ mc admin info myminio
> ●  localhost:9000
>    Uptime : online since 6 minutes ago
>   Version : 2018-01-10T17:54:10Z
>    Region : 
>  SQS ARNs : arn:minio:sqs::test2:redis arn:minio:sqs::test1:redis 
>   Network : Incoming 10KiB, Outgoing 7.8KiB
>   Storage : Total 156GiB, Free 75GiB
> 
> ```
> localhost:9000 is wrong in this.

It is the current behaviour (i.e. same without this PR). What should it be if not localhost:9000? If it is not related to this PR, I can fix it in existing mc/server.

> Bucket is not valid is perhaps not the correct message here, we need to guide the users here.
> 
> ```
> mc admin lock list myminio
> mc: <ERROR> Cannot get lock status. The specified bucket is not valid.
> 
> ```

The problem here is that the bucket is not specified. I will send a separate PR adding validation to fix this.

File: cmd/admin-heal.go [hunk 4]
```diff
@@ -72,304 +76,556 @@ EXAMPLES:
...
+	colRow := hColTable[parityShards]
+	for index, val := range colRow {
+		if val != -1 && surplusShards <= val {
+			return hColOrder[index], err
+		}
+	}
+	return c, fmt.Errorf("cannot get a heal color code")
+}
+
+type hri struct {
```
Comment (vadmeste) [inline cmd/admin-heal.go]: Can you move HRI related code to a new source file ?

File: cmd/admin-heal.go [hunk 5]
```diff
@@ -72,304 +76,556 @@ EXAMPLES:
...
-	case madmin.CanHeal:
-		msg += "can be healed"
-	case madmin.Corrupted:
-		msg += "cannot be healed"
-	case madmin.QuorumUnavailable:
-		msg += "quorum not available for healing"
-	}
-	msg += ".\n"
-	return console.Colorize("Heal", msg)
+type uiData struct {
```
Comment (vadmeste) [inline cmd/admin-heal.go]: Can you move uiData code to a new source file ?

Comment (donatello) [reply cmd/admin-heal.go]: Done.

File: cmd/admin-heal.go (hunk 4 above)
Comment (donatello) [reply cmd/admin-heal.go]: Done.

File: cmd/admin-heal.go [hunk 6]
```diff
@@ -19,37 +19,41 @@ package cmd
...
+	cli.BoolFlag{
+		Name:  "recursive, r",
+		Usage: "Heal recursively",
+	},
+	cli.BoolFlag{
+		Name:  "dry-run, n",
+		Usage: "Only inspect data, but do not mutate",
+	},
+	cli.BoolFlag{
+		Name:  "incomplete, I",
```
[the remark this answers is not in the document]
Comment (donatello) [reply cmd/admin-heal.go]: Done.

File: cmd/admin-service-stop.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
+/*
+ * Minio Client (C) 2017 Minio, Inc.
```
[the remark this answers is not in the document]
Comment (donatello) [reply cmd/admin-service-stop.go]: Done.

**Comment by vadmeste** [comment]
@donatello, I think you can still move getHPrintCol, getHColCode and lineTrunc to another file, the one or more which depend on these functions.

File: cmd/admin-heal.go [hunk 8]
```diff
@@ -29,23 +27,24 @@ import (
...
-		},
-	}
-)
+var adminHealFlags = []cli.Flag{
+	cli.BoolFlag{
+		Name:  "recursive, r",
+		Usage: "Heal recursively",
+	},
+	cli.BoolFlag{
+		Name:  "dry-run, n",
```
Comment (vadmeste) [inline cmd/admin-heal.go]: Examples are not updated, --fake is still being used.

**Comment by vadmeste** [comment]
```
18:08 $ mc admin heal myminio/testbucket/1mb.bin
 ◓  testbucket/1mb.bin
    0/1 objects; 1024 KiB in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 3 │ 100.0% ████████████ │
    │ Yellow │ 0 │   0.0%              │
    │ Red    │ 0 │   0.0%              │
    │ Grey   │ 0 │   0.0%              │
    └────────┴───┴─────────────────────┘
```

I am not sure I understand exactly '0/1 objects', or it has to be '1/1 objects' ?

File: cmd/admin-heal.go (hunk 8 above)
Comment (donatello) [reply cmd/admin-heal.go]: Fixed.

File: cmd/admin-heal-result-item.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+		surplus := numAvail - quorum
+		parity := h.DiskCount - quorum
+		c, err = getHColCode(surplus, parity)
+		return
+	}
+
+	onlineBefore, onlineAfter := h.GetOnlineCounts()
+	b, err = getColCode(onlineBefore)
+	if err != nil {
+		return
```
[the remark this answers is not in the document]
Comment (donatello) [reply cmd/admin-heal-result-item.go]: These are system metadata files that are healed such as format.json, and bucket metadata files such as policy.json, notifications.xml, etc.

**Comment by donatello** [comment]
@vadmeste 
>  I think you can still move getHPrintCol, getHColCode and lineTrunc to another file, the one or more which depend on these functions.

Done.

> I am not sure I understand exactly '0/1 objects', or it has to be '1/1 objects' ?

The phrase "1/3 objects" implies that 1 object was healed out of 3 objects scanned. See the `-q` and `--json` options too.

**Comment by donatello** [comment]
This PR is building again.

**Pull request merged**